### PR TITLE
FIR-25300 improved discovery of isReadOnly() for connection and datab…

### PR DESCRIPTION
**Pull request by alexradzin**
…ase metadata.

### CODE REVIEW

File: src/main/java/com/firebolt/jdbc/metadata/MetadataUtil.java [hunk 1]
```diff
@@ -102,4 +102,14 @@ public static String getDatabaseVersionQuery(String engine) {
 				.conditions(Collections.singletonList(String.format("engine_name iLIKE '%s%%'", engine))).build()
 				.toSql();
 	}
+
+	// lower() is used in SQL statement because the comparison in Firebolt is case-sensitive but we retrieve engine name from host where lowercase is used.
+	// But default engine is created with suffix Ingest (upper I). This is the reason to call loser() function.
```
Comment (ptiurin) [inline src/main/java/com/firebolt/jdbc/metadata/MetadataUtil.java:107]: typo: `lower`

File: src/main/java/com/firebolt/jdbc/metadata/MetadataUtil.java [hunk 2]
```diff
@@ -102,4 +102,14 @@ public static String getDatabaseVersionQuery(String engine) {
...
+	public static String isConnectionReadOnly() {
+		return "select count(*) = 0 from information_schema.engines where status = 'Running' and engine_type = 'General Purpose' and lower(engine_name) = lower(?)";
+	}
+
+	public static String isDatabaseReadOnly() {
+		return "select count(*) = 0 from information_schema.engines where status = 'Running' and engine_type = 'General Purpose' and lower(attached_to) = lower(?)";
+	}
```
Comment (ptiurin) [inline src/main/java/com/firebolt/jdbc/metadata/MetadataUtil.java:114]: This is fine for Firebolt 1.0, but just a heads-up - FB 2.0 will have RBAC so cases when a user doesn't have access to GP engine or user is set to reader despite working on a GP engine will be possible.

File: src/integrationTest/java/integration/tests/IsReadableTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            log.info("end of the engines list");
+
+            PreparedStatement ps2 = connection.prepareStatement("select count(*) = 0 from information_schema.engines where status = 'Running' and engine_type = 'General Purpose' and lower(attached_to) = lower(?)");
+            ps2.setString(1, database);
+            ResultSet rs2 = ps2.executeQuery();
+            while (rs2.next()) {
+                boolean isReadOnly = rs2.getBoolean(1);
+                log.info("read only={}", isReadOnly);
+            }
+            log.info("end of the readonly");
```
Comment (ptiurin) [inline src/integrationTest/java/integration/tests/IsReadableTest.java:41]: Why are we doing those statements in a test? Is this leftover from debugging?

File: src/test/java/com/firebolt/jdbc/metadata/FireboltDatabaseMetadataTest.java [hunk 4]
```diff
@@ -355,7 +356,35 @@ void shouldGetDatabaseMinorVersion() throws SQLException {
...
+//		when(fireboltConnection.prepareStatement(anyString())).thenReturn(ps);
+//		when(ps.executeQuery()).thenReturn(rs);
+//		when(rs.next()).thenReturn(false);
+//		assertFalse(fireboltDatabaseMetadata.isReadOnly());
+
+
+//		ResultSet rs  = mock(ResultSet.class);
+//		when(rs.next()).thenReturn(false);
+//		isReadOnly(rs, false);
+	}
```
Comment (ptiurin) [inline src/test/java/com/firebolt/jdbc/metadata/FireboltDatabaseMetadataTest.java:370]: comments

File: src/integrationTest/java/integration/tests/IsReadableTest.java (hunk 3 above)
Comment (alexradzin) [reply src/integrationTest/java/integration/tests/IsReadableTest.java]: no, this code is added for debugging.

File: src/main/java/com/firebolt/jdbc/metadata/MetadataUtil.java (hunk 1 above)
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/metadata/MetadataUtil.java:107]: 10x

File: src/main/java/com/firebolt/jdbc/metadata/MetadataUtil.java (hunk 2 above)
Comment (alexradzin) [reply src/main/java/com/firebolt/jdbc/metadata/MetadataUtil.java:114]: Well... Do you have better solution? I will appreciate any help.

File: src/integrationTest/java/integration/tests/IsReadableTest.java (hunk 3 above)
Comment (alexradzin) [reply src/integrationTest/java/integration/tests/IsReadableTest.java]: removed redundant code

File: src/main/java/com/firebolt/jdbc/metadata/MetadataUtil.java (hunk 2 above)
Comment (ptiurin) [reply src/main/java/com/firebolt/jdbc/metadata/MetadataUtil.java]: Don't know. You'll need to check with the relevant team if there's a way to verify access. #sql-frontent will probably be a good place to start. This is the only driver that exposes `is_ReadOnly` so we haven't had the same issue elsewhere.
Now that V1 and V2 are merged `is_ReadOnly` will potentially return false info with RBAC so I'm not sure if we should implement it in the current state.

**Pull request closed without merging**